### ISSUE TITLE
HLS infinite buffering on concrete link

### ISSUE THREAD

**Issue by personshelldon** [google/ExoPlayer#10954]
### ExoPlayer Version

2.18.2

### Devices that reproduce the issue

Android TV Emulator API 28
Android TV Emulator API 33
Nvidia Shield TV (Android 10)
Sony Bravia 4K (Android 11)
MAG500 (Android 10)
MAG425A (Android 9)
OnePlus 10 PRO (Android 13)
OnePlus 8T (Android 12)

### Reproducible in the demo app?

Yes

### Reproduction steps

1. Add the attached link to the demo app and force the media type of HLS
2. Play the link

### Expected result

The video plays after buffering OR video plays without sound OR the error because of MPEG2 audio tracks

### Actual result

Infinite buffering in the custom app and error after long buffering in the demo app.

### Media

https://edge.simplware.pro/eurosport/index.m3u8?token=mud6TFfRrDO9cMORqGJJ5060fgHxS0UCU8ubKTeU6nLDseVyDe

### Bug Report

- [ ] You will email the zip file produced by `adb bugreport` to dev.exoplayer@gmail.com after filing this issue.

**Comment by personshelldon** [comment]
Is there any update on this problem? Thank You.

**Comment by tianyif** [comment]
Hi @personshelldon,

Thanks for reporting! I couldn't reproduce the infinite buffering with the media url on Sony Bravia 4K device. Also, could you please clarify the actual and expected results? In the "actual result" section, you mentioned there was an error after the long buffering, could you please **add the stack trace of the error**? And I'm a bit confused by the "expected result" section (which says " ... OR video plays without sound OR the error because of MPEG2 audio tracks"), are you expecting some errors with the audio?

**Comment by personshelldon** [comment]
> Thanks for reporting! I couldn't reproduce the infinite buffering with the media url on Sony Bravia 4K device

Hello! The problem is that that no errors at all, but the video is not going to play. The last event that I receive was `STATE_BUFFERING`. After that nothing happens until I press pause or stop. I'm expecting some errors with the audio because I do not know if the devices can support this audio format.

But the problem is that video won't play at all without any errors. Also, I tested this video on different platforms and the problem can be reproduced on Android and iOS (iOS also do not produce any errors, but only the first frame is shown with audio playback in background). Roku and WebOS can play this video correctly. Also, VLC can play this video successfully on all platforms including Android but not ExoPlayer.

What logs I can prepare for You?

**Comment by personshelldon** [comment]
Hello! Is there any update on this issue? Thank You in advance!

**Comment by tianyif** [comment]
Hi @personshelldon,

Sorry for the late response! I haven't been able to reproduce this issue with the provided link, but yesterday when I looked another issue https://github.com/google/ExoPlayer/issues/11041, I was thinking the scenario they faced was very closed to what you described:

* The playback state stays at `STATE_BUFFERING` without triggering STATE_READY`. Then video doesn't play.
* No errors thrown.
* VLC player can play successfully.

The tl;dr of that another issue is that the stream doesn't contain IDR frames, which makes the `MediaCodecVideoRenderer` never get ready. The workaround is set the FLAG_ALLOW_NON_IDR_KEYFRAMES in the [DefaultTsPayloadReaderFactory](https://github.com/google/ExoPlayer/blob/r2.18.4/library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/DefaultTsPayloadReaderFactory.java#L116).

When I tested on your live stream today, I was able to see the the stream contains `NalUnitUtil.NAL_UNIT_TYPE_IDR`, so that's why I didn't see the described issue now. And when you saw the issue earlier, that may be the case.

Sorry that I was not be able to provide a concrete answer to your problem, but hopefully the above can give some hints on what was happening before.

**Comment by personshelldon** [comment]
Hello @tianyif! Thank You for your reply! I have some time to test your solution and the problem definitely not in IDR frames. Also, I made some tests with different ExoPlayer versions and what I found:

- For now, my app has ExoPlayer version 2.17.1. I downloaded ExoPlayer Demo v2.17.1 from GitHub and tested my link in it. 
The result - is infinite buffering as in my app.
- I made the same test with ExoPlayer 2.18.4 and the video started to play successfully but the error occurs very fast: sometimes only the first frame is displayed and then an error or the player can play only the first chunk of the video. Also, the player successfully reported that audio formats are not supported.

In the second case, the error is always the same  on the emulator (can be easily reproduced): 
```
2023-03-15 21:08:41.331 22177-22351 CCodec                  com.google.android.exoplayer2.demo   W  [c2.goldfish.h264.decoder] previous call to queue exceeded timeout
2023-03-15 21:08:41.331 22177-24394 MediaCodec              com.google.android.exoplayer2.demo   E  Codec reported err 0x80000000/UNKNOWN_ERROR, actionCode 0, while in state 6/STARTED
2023-03-15 21:08:41.331 22177-24394 SurfaceUtils            com.google.android.exoplayer2.demo   D  disconnecting from surface 0x73379ae282b0, reason disconnectFromSurface
2023-03-15 21:08:41.383 22177-22354 BufferQueueProducer     com.google.android.exoplayer2.demo   E  [SurfaceView[com.google.android.exoplayer2.demo/com.google.android.exoplayer2.demo.PlayerActivity]#11(BLAST Consumer)11](id:56a10000000b,api:0,p:-1,c:22177) cancelBuffer: BufferQueue has no connected producer
2023-03-15 21:08:41.389 22177-24422 hw-BpHwBinder           com.google.android.exoplayer2.demo   I  onLastStrongRef automatically unlinking death recipients
2023-03-15 21:08:41.397 22177-24381 MediaCodecVideoRenderer com.google.android.exoplayer2.demo   E  Video codec error
                                                                                                      android.media.MediaCodec$CodecException: Error 0x80000000
2023-03-15 21:08:41.398 22177-24381 ExoPlayerImplInternal   com.google.android.exoplayer2.demo   E  Playback error
                                                                                                      com.google.android.exoplayer2.ExoPlaybackException: MediaCodecVideoRenderer error, index=0, format=Format(0, null, null, video/avc, avc1.4D4028, 7600000, null, [1920, 1080, 25.0], [-1, -1]), format_supported=YES
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:578)
                                                                                                          at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                                          at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                          at android.os.Looper.loop(Looper.java:288)
                                                                                                          at android.os.HandlerThread.run(HandlerThread.java:67)
                                                                                                      Caused by: com.google.android.exoplayer2.video.MediaCodecVideoDecoderException: Decoder failed: c2.goldfish.h264.decoder
                                                                                                          at com.google.android.exoplayer2.video.MediaCodecVideoRenderer.createDecoderException(MediaCodecVideoRenderer.java:1639)
                                                                                                          at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:794)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:1007)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:502)
                                                                                                          at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                                          at android.os.Looper.loopOnce(Looper.java:201) 
                                                                                                          at android.os.Looper.loop(Looper.java:288) 
                                                                                                          at android.os.HandlerThread.run(HandlerThread.java:67) 
                                                                                                      Caused by: android.media.MediaCodec$CodecException: Error 0x80000000
2023-03-15 21:08:41.404 22177-22177 EventLogger             com.google.android.exoplayer2.demo   D  videoDisabled [eventTime=7.04, mediaPos=1.13, window=0, period=0]
2023-03-15 21:08:41.404 22177-24381 ExoPlayerImplInternal   com.google.android.exoplayer2.demo   E  Disable failed.
                                                                                                      java.lang.IllegalStateException
                                                                                                          at android.media.MediaCodec.native_flush(Native Method)
                                                                                                          at android.media.MediaCodec.flush(MediaCodec.java:2387)
                                                                                                          at com.google.android.exoplayer2.mediacodec.AsynchronousMediaCodecAdapter.flush(AsynchronousMediaCodecAdapter.java:230)
                                                                                                          at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.flushCodec(MediaCodecRenderer.java:858)
                                                                                                          at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.flushOrReleaseCodec(MediaCodecRenderer.java:851)
                                                                                                          at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.onDisabled(MediaCodecRenderer.java:691)
                                                                                                          at com.google.android.exoplayer2.video.MediaCodecVideoRenderer.onDisabled(MediaCodecVideoRenderer.java:609)
                                                                                                          at com.google.android.exoplayer2.BaseRenderer.disable(BaseRenderer.java:186)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.disableRenderer(ExoPlayerImplInternal.java:1670)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.resetInternal(ExoPlayerImplInternal.java:1411)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.stopInternal(ExoPlayerImplInternal.java:1372)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:594)
                                                                                                          at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                                          at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                          at android.os.Looper.loop(Looper.java:288)
                                                                                                          at android.os.HandlerThread.run(HandlerThread.java:67)
2023-03-15 21:08:41.406 22177-22177 EventLogger             com.google.android.exoplayer2.demo   D  videoDecoderReleased [eventTime=7.04, mediaPos=1.13, window=0, period=0, c2.goldfish.h264.decoder]
2023-03-15 21:08:41.408 22177-22177 EventLogger             com.google.android.exoplayer2.demo   E  playerFailed [eventTime=7.04, mediaPos=1.13, window=0, period=0, errorCode=ERROR_CODE_DECODING_FAILED
                                                                                                      com.google.android.exoplayer2.ExoPlaybackException: MediaCodecVideoRenderer error, index=0, format=Format(0, null, null, video/avc, avc1.4D4028, 7600000, null, [1920, 1080, 25.0], [-1, -1]), format_supported=YES
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:578)
                                                                                                          at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                                          at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                          at android.os.Looper.loop(Looper.java:288)
                                                                                                          at android.os.HandlerThread.run(HandlerThread.java:67)
                                                                                                      Caused by: com.google.android.exoplayer2.video.MediaCodecVideoDecoderException: Decoder failed: c2.goldfish.h264.decoder
                                                                                                          at com.google.android.exoplayer2.video.MediaCodecVideoRenderer.createDecoderException(MediaCodecVideoRenderer.java:1639)
                                                                                                          at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:794)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:1007)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:502)
                                                                                                          at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                                          at android.os.Looper.loopOnce(Looper.java:201) 
                                                                                                          at android.os.Looper.loop(Looper.java:288) 
                                                                                                          at android.os.HandlerThread.run(HandlerThread.java:67) 
                                                                                                      Caused by: android.media.MediaCodec$CodecException: Error 0x80000000
                                                                                                    ]
```

But I tried the second solution on the real device and the video plays correctly for some time. I will add a new ExoPlayer to my project and test it on different real devices and will report the results after that. Thank You!

So, the problem was fixed (maybe partially) something between versions of the ExoPlayer 2.17.1 and 2.18.4.

**Comment by tianyif** [comment]
Thanks @personshelldon so much for the really helpful information! I further narrowed down the versions and thought https://github.com/google/ExoPlayer/commit/b7cf772afbc7eb8997d9cfb2938aab53ee7a58ea could be the one that fixed your infinite buffering issue.

We enabled HLS [chunkless preparation](https://medium.com/google-exoplayer/faster-hls-preparation-f6611aa15ea6) by default and this bug was around 2.17.1 and we fixed them starting from 2.18.0. Without that fix, I was able to see the preparation of your stream will be [waiting](https://github.com/google/ExoPlayer/blob/r2.17.1/library/common/src/main/java/com/google/android/exoplayer2/util/TimestampAdjuster.java#L112) forever, without further proceeding.

So you can either upgrade your ExoPlayer version and follow the [workaround](https://b.corp.google.com/issues/233426489#comment6) to disable the chunkless preparation for 2.17.1 ExoPlayer. 

I noticed that you initially reported this issue for 2.18.2. Based on the above theory and my debugging experience of not being able to reproduce, the issue should be no longer there on 2.18.2. Could you please help me confirm if the reported version should be corrected? Just hope to keep an easier reference for the other issue reporters and triagers :)

**Comment by personshelldon** [comment]
Hello @tianyif! I've tested [b7cf772](https://github.com/google/ExoPlayer/commit/b7cf772afbc7eb8997d9cfb2938aab53ee7a58ea) and this really helped on 2.17.1. Also, I've tested this solution with FFMpeg extension (FFMpeg 5.1.2) and now the video can be played correctly.

Tested on devices from the first message: correctly works on real devices, but in an emulator (AndroidTV) always a playback error within 1-10 seconds of the playback (the error in the message above). 

Because on real devices this fix has helped, I think the task can be closed! Thank You!
